Update SDA Downscaling to allow sf

dlebauer opened the pull request:
## Description

* Refactored SDA functions to accept direct R objects (lists, data.frames, and sf objects) in addition to file paths.
* Added a helper function (.convert_coords_to_sf()) to standardize coordinate conversion.
* Updated man pages to reflect changes in input parameters.
* Implemented new tests to verify the updated behavior.

 

## Motivation and Context

Existing downscaling code required rasters. In the CCMMF project we are using vector data, so this PR generalizes the downscaling code to support vector file types as well.

It also allows ensemble data to be passed as either an RDS file or an R list of dataframes.

## Review Time Estimate

- [x] Without unnecessary delay

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [x] My name is in the list of CITATION.cff
- [x] I agree that PEcAn Project may distribute my contribution under any or all of
	- the same license as the existing code,
	- and/or the BSD 3-clause license.
- [x] I have updated the CHANGELOG.md.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

dlebauer committed:
also accept sf pts object

dlebauer committed:
update CHANGELOG.md

dlebauer committed:
changed SDA downscale argument names to match new functionality

dlebauer committed:
make document

dlebauer committed:
merge upstream develop; resolve changelog conflict

dlebauer commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -114,62 +153,70 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
+    # For raster objects, convert the sf coordinates to SpatVector before extraction
+    predictors <- as.data.frame(terra::extract(covariates, terra::vect(site_coordinates_sf), ID = FALSE))
+  } else if (inherits(covariates, "sf")) {
+    # For sf objects, perform a spatial join using st_intersects and then drop the geometry column
```
```suggestion
    # For sf objects, spatial join
```

dlebauer committed:
Update modules/assim.sequential/R/downscale_function.R

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -1,67 +1,104 @@
+##'   - A file path to a `.csv` containing site coordinates with columns `"id"`, `"lat"`, and `"lon"`.
+##'   - A `data.frame` with the same structure.
+##'   - An `sf` object with point geometries.
+##' @param date Date. If SDA site run, format is yyyy/mm/dd; if NEON, yyyy-mm-dd. Restricted to years within the file or object supplied to 'ensemble_data'.
```
When this function was originally written I asked that this be fixed to just take in an actual Date. While you're fixing things could you fix that too.

dlebauer replied:
@mdietze I'm not sure if I fully understand, and how I can make sure not to brake existing code. 

What I have in mind is to 
- require the date parameter be a date object?
- if it is not a date object, try to convert from the two yyyymmdd options?

is that what you had in mind?

dlebauer replied:
@mdietze I'm not sure if I fully understand, and how I can make sure not to brake existing code. 

What I have in mind is to 
- require the date parameter be a Date object and update the docs accordingly and
- if it is not a date object, try to convert from the two yyyymmdd options? (already happens)

is that what you had in mind?

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -1,67 +1,104 @@
-  
+
+  # If sf object, convert to data.frame with 'lon' and 'lat'
+  if (inherits(site_coordinates, "sf")) {
```
Curious why this is down here rather than up earlier where the other two site_coordinates cases were

dlebauer replied:
good point, no real reason

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -82,28 +119,30 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
-##' @param date Date. If SDA site run, format is yyyy/mm/dd; if NEON, yyyy-mm-dd. Restricted to years within file supplied to 'preprocessed' from the 'data_path'.
-##' @param carbon_pool Character. Carbon pool of interest. Name must match carbon pool name found within file supplied to 'preprocessed' from the 'data_path'.
-##' @param covariates SpatRaster stack. Used as predictors in downscaling. Layers within stack should be named. Recommended that this stack be generated using 'covariates' instructions in assim.sequential/inst folder
+##' @param date Date. If SDA site run, format is yyyy/mm/dd; if NEON, yyyy-mm-dd. Restricted to years within file supplied to 'preprocessed' from the 'ensemble_data'.
```
same comment about fixing date

dlebauer replied:
In this case, it doesn't look like the function actually uses date, so I've removed it

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -114,62 +153,70 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
-                                                keep.forest = TRUE,
-                                                importance = TRUE)
-      
+        data = train_data,
```
revert indentation change

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -114,62 +153,70 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
+        keep.forest = TRUE,
+        importance = TRUE
+      )
+
```
Another suggestion I've made in the past is that there may be logical/computational advantages to separating the function that builds the downscaling models and a separate function that makes the predictions to a map. Right now one can't just fit the model and evaluate the results without also running the (more computationally costly) mapping

dlebauer replied:
good idea, though this will be nontrivial refactor; will add as TODO 

```suggestion
# TODO: Separate model fitting and held-out evaluation from spatial map prediction,
# so models can be evaluated without the cost of generating raster maps.
```

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -302,18 +349,19 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
-                                  fun = cnn_ensemble_predict,
-                                  scaling_params = scaling_params)
+      maps[[i]] <- terra::predict(prediction_rast,
+        model = models[[i]],
```
fix indent

infotroph commented on modules/assim.sequential/tests/testthat/test_downscaling.R:
```diff
@@ -0,0 +1,79 @@
+library(testthat)
+library(sf)
+library(terra)
+library(randomForest)
+library(withr)
```
Consider removing these -- testthat gets loaded by the package-level testthat.R, and since we want package functions to behave right when a dependency isn't attached, I think there's virtue in making ourselves run the tests without attaching things.

dlebauer committed:
SDA_downscale_preprocess now expects Date objects,
Deprecating date arg to SDA_downscale because it is never used
put check for sf object with related checks

dlebauer committed:
update SDA downscale docs

infotroph commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -112,64 +153,85 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
+    #    1. line starting with ##' @param date
+    #    2. date = NULL from function call
+    #    3. this conditional starting with if (!missing(date)) and ending after the warning below
+    PEcAn.logger::logger.warn("'date' argument is not used and will be removed on or after 2026-04-01. It is currently ignored.", call. = FALSE)
```
logger.warn doesn't take `call.` and treats its contents as another message to print
```suggestion
    PEcAn.logger::logger.warn("'date' argument is not used and will be removed on or after 2026-04-01. It is currently ignored.")
```

(Also `call.` is high up my list of the base R arguments that look the most like typos, but that's not our problem to fix 😉 )

infotroph reviewed (commented):
A high-level question that you likely considered already: 

Accepting both files and in-memory data seems like it adds a lot of complexity to functions that were nice and clean before that, and I naively expect it would be both "easy" and desirable for error recovery to instead write data out to files before running this. What tradeoff factors make this the better approach?

mdietze commented:
Per @infotroph last comment, I'm inclined to push in the other direction. For SDA_downscale, I'm fairly confident that I asked the original authors of these functions to stop loading the data within the function, as you have greater flexibility if you load the constraint data first then call the function. This way you can easily control what covariate variables are part of any particular downscaling, and what domain you're doing the downscaling for, by changing what stack of variables you're passing in. In other words, for that function you could probably drop the option to pass in file paths.

infotroph commented:
@mdietze 's point makes sense and yes, load-first seems sensible. I was trying to question the need to support both rather than pushing for write-first specifically.

dlebauer committed:
remove deprecated date handling

dlebauer commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -1,67 +1,106 @@
     } else {
       message("Truncating carbon data to match site coordinates rows.")
-      carbon_data <- carbon_data[1:nrow(site_coordinates), ]
+      carbon_data <- carbon_data[seq_len(nrow(site_coordinates)), ]
```
@sambhavnoobcoder it looks like you implemented this, could you help me understand the logic? My suspicion is that if `nrow(site_coordinates) != nrow(carbon_data)`, something wrong with the inputs, and truncating one or the other could result in carbon_data being associated with the wrong sites.

sambhavnoobcoder replied:
Hi @dlebauer ,my original implementation assumed that the site coordinates and carbon data were already aligned (i.e., that the order of rows in both datasets corresponds one‐to‐one). The idea was that if one of the inputs had extra rows, then simply truncating the longer dataset would let the function run without error. In that context, the truncation was a quick fix for handling extra or misordered rows.

That said, the concern is valid: if the rows aren’t truly aligned, then truncating without verifying a matching site identifier could indeed lead to carbon data being paired with the wrong sites. In other words, the current approach works correctly only if the inputs are pre‐validated or guaranteed to be in the correct order.

dlebauer committed:
Update modules/assim.sequential/R/downscale_function.R

Co-authored-by: Chris Black <chris@ckblack.org>

dlebauer commented:
> @mdietze 's point makes sense and yes, load-first seems sensible. I was trying to question the need to support both rather than pushing for write-first specifically.

I was maintaining the 'read from file' functionality to remain backwards-compatible. If these functions don't need to be backward compatible, they will be much easier to refactor.

dlebauer committed:
Refactoring and debugging SDA_downscale_preprocess; update test cases

mdietze commented:
Right now, this code has a small enough user community that backwards compatibility isn't an issue. That will change once we start publishing using this workflow and need outward-facing reproducibility

dlebauer committed:
merge

dlebauer committed:
Merge branch 'develop' into ensemble_downscaling

dlebauer committed:
add refactored ensemble_downscale function derived from SDA_downscale but doesn't use raster covariates

dlebauer committed:
reverted original downscale_funciton.R so it doesn't conflict with #3451

dlebauer committed:
Merge branch 'ensemble_downscaling' of github.com:dlebauer/pecan into ensemble_downscaling

dlebauer commented on CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@ For more information about this file see also [Keep a Changelog](http://keepacha
 ### Fixed
 - updated github action to build docker images
 - PEcAn.SIPNET now accepts relative paths in its input XML (#3418). Previously all files referenced in the autogenerated `job.sh` needed to be specified as absolute paths.
+- Refactored `SDA_downscale` and `SDA_downscale_preprocess` to allow direct use of R objects in addition to file paths. Added tests for these functions.
```
```suggestion
- Refactored and created new version of `SDA_downscale` named `ensemble_downscale`.
```

dlebauer committed:
Update CHANGELOG.md

dlebauer committed:
restore refactored SDA_downscale function

dlebauer committed:
Merge branch 'ensemble_downscaling' of github.com:dlebauer/pecan into ensemble_downscaling

dlebauer committed:
fix borked find and replace

dlebauer committed:
update documentation

dlebauer committed:
revised ensemble downscale

dlebauer committed:
export subset_ensemble and ensemble_downscale functions

dlebauer committed:
update metrics function

dlebauer committed:
Merge branch 'develop' of github.com:pecanproject/pecan into ensemble_downscaling

dlebauer committed:
explicitly export outlier.etector.boxplot so it isn't interpreted as an S3 method

infotroph commented on CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@ For more information about this file see also [Keep a Changelog](http://keepacha
 - updated github action to build docker images
 - PEcAn.SIPNET now accepts relative paths in its input XML (#3418). Previously all files referenced in the autogenerated `job.sh` needed to be specified as absolute paths.
+- Refactored and created new version of `SDA_downscale` named `ensemble_downscale`.
+- Added helper function `.convert_coords_to_sf()` for consistent conversion of data with lat lon data to `sf` pts.
```
Nits: 
* refactors seem like belong in "changed" not "fixed"
* internal helper functions don't need to be mentioned in the changelog (and I'd argue it's confusing to do so, since the point is for users to not know or care about them)

infotroph replied:
Also: Need to move up to "unlreased" (this section is now for already-tagged 1.9.0)

infotroph commented on modules/assim.sequential/R/Helper.functions.R:
```diff
@@ -6,10 +6,10 @@
 #' @return A list the same dimension as X, with each column of each dataframe
 #'   modified by replacing outlier points with the column median
-#' @export
+#' @export outlier.detector.boxplot
```
I bet this is trying to prevent this name being treated as an S3 method, but I recommend renaming the function instead. Life's too short to fight with R about this.

dlebauer replied:
renamed replacing `.` w/ `_`

infotroph commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -1,67 +1,113 @@
+  # Ensure 'date' is a Date object, if not, convert
+  if (!inherits(date, "Date")) {
+    standard_date <- lubridate::ymd(date)
+  }
```
`ymd` accepts Date objects and returns them unchanged -- can skip the if and the rename

infotroph commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -1,67 +1,113 @@
+  #    } else {
+  #      PEcAn.logger::logger.info("Truncating carbon data to match site coordinates rows.")
+  #      carbon_data <- carbon_data[seq_len(nrow(site_coordinates)), ]
+  #    }
```
The link is taking me to a whole-PR view so I can't tell what specific comment/commit it's referring to, but the commented lines seem safe to delete -- I agree that unequal row counts should make us stop with an error rather than truncate and continue.

infotroph commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -82,28 +128,30 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
-##' @param date Date. If SDA site run, format is yyyy/mm/dd; if NEON, yyyy-mm-dd. Restricted to years within file supplied to 'preprocessed' from the 'data_path'.
-##' @param carbon_pool Character. Carbon pool of interest. Name must match carbon pool name found within file supplied to 'preprocessed' from the 'data_path'.
-##' @param covariates SpatRaster stack. Used as predictors in downscaling. Layers within stack should be named. Recommended that this stack be generated using 'covariates' instructions in assim.sequential/inst folder
+##' @param date *Deprecated*. This argument has never been used and will be removed after 2026-04-01
```
I vote drop it now, mention in NEWS, and let users figure it out

infotroph commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -112,64 +160,82 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
+  # Convert site coordinates to an sf object using the helper function
+  site_coordinates_sf <- .convert_coords_to_sf(preprocessed$site_coordinates)
+
+  # Extract predictors from covariates <U+2013> support both SpatRaster and sf objects:
```
`<U+2013>`?

infotroph commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -112,64 +160,82 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
+  # Extract predictors from covariates <U+2013> support both SpatRaster and sf objects:
+  if (inherits(covariates, "SpatRaster")) {
+    # For raster objects, convert the sf coordinates to SpatVector before extraction
+    predictors <- as.data.frame(terra::extract(covariates, terra::vect(site_coordinates_sf), ID = FALSE))
```
Comment and code mismatch -- Doesn't the `as.data.frame` mean this converts to plain df rather than SpatVector?

infotroph commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -302,18 +368,19 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
     }
   } else {
-    stop("Invalid model_type. Please choose either 'rf' for Random Forest or 'cnn' for Convolutional Neural Network.")
+    PEcAn.logger::logger.error("Invalid model_type. Please choose either 'rf' for Random Forest or 'cnn' for Convolutional Neural Network.")
```
I know you're just changing stop to logger.error here, but boy is this else a long way from the other cases. A nice refactor would be to do the argument validation with `match.arg(model_type)` way up at the top of the function rather than needing to come back to it at the very bottom.

infotroph replied:
I might even go one step further and pull the if cases out to helper functions so that the entire control flow could be 
```
SDA_downscale <- function ([...], model_type = c("rf", "cnn"), [...]) {
  model_type <- match_arg(model_type)
  [...]
  if (model_type == "rf") {
    predictions <- downscale_rf([...])
  } else {
    predictions <- downscale_cnn([...])
  }
[...]
```

dlebauer replied:
Implemented match.arg() validation at top of SDA_downscale. Will leave extracting the RF and CNN branches into helpers for future refactor. 

Added TODO w/ suggestion about refactoring branches into helper functions

infotroph commented on modules/assim.sequential/R/ensemble_downscale.R:
```diff
@@ -0,0 +1,266 @@
+}
+
+## Helper function to convert table with lat, lon into an sf object
+.convert_coords_to_sf <- function(coords) {
```
Same function name used here and in downscale_function.R. Only whichever definition is parsed last at package byte-compilation time will be used (which I think is determined by collation order). Given Helper.functions.R already exists, I recommend moving all these dotted-name functions there.

dlebauer replied:
removed here

infotroph commented on modules/assim.sequential/R/ensemble_downscale.R:
```diff
@@ -0,0 +1,266 @@
+
+  # Filter the ensemble data to the specified date and carbon pool
+  ensemble_data <- ensemble_data |>
+    filter(
```
```suggestion
    dplyr::filter(
```

infotroph replied:
you'll probably need `.data$` on the column names here too

infotroph replied:
Same for other tidyverse fn calls below

infotroph commented on modules/assim.sequential/R/ensemble_downscale.R:
```diff
@@ -0,0 +1,266 @@
+      site_id %in% unique(site_coords$id),
+      variable == carbon_pool
+    ) |>
+    select(site_id, ensemble, prediction)  # use site_id instead of site
```
```suggestion
    dplyr::select("site_id", "ensemble", "prediction")  # use site_id instead of site
```
not sure what the comment means -- is it needed?

infotroph commented on modules/assim.sequential/R/ensemble_downscale.R:
```diff
@@ -0,0 +1,266 @@
+
+  ensembles <- unique(ensemble_data$ensemble)
+
+  results <- furrr::future_map(seq_along(ensembles), function(i) {
```
I strongly encourage pulling this anonymous function out to its own definition and giving it an informative name

infotroph replied:
Also it appears `i` is only used to subset `ensemble` -- is there a reason not to pass each one directlly (i.e `map(ensembles, function(ens) ...`?

infotroph commented on modules/assim.sequential/R/ensemble_downscale.R:
```diff
@@ -0,0 +1,266 @@
+##' @export
+downscale_metrics <- function(downscale_output) {
+  
+  test_data_list <- lapply(downscale_output$test_data, function(x) dplyr::pull(x, prediction))
```
curious why the extra lambda -- does this differ from `lapply(downscale_output$test_data, dplyr::pull, prediction)`, or indeed from ```lapply(downscale_output$test_data, `[[`, "prediction")```?

dlebauer replied:
implemented [[ approach

infotroph commented on modules/assim.sequential/tests/testthat/test_downscaling.R:
```diff
@@ -0,0 +1,82 @@
+withr::with_tempdir({
+  temp_ensemble_data_rds <- "ensemble_data.rds"
+  temp_coords_csv <- "final_design_points.csv"
+  file.remove(temp_ensemble_data_rds, temp_coords_csv)
```
Why do you need to start by removing these if the whole test happens in a fresh tempdir?

infotroph commented on modules/assim.sequential/tests/testthat/test_downscaling.R:
```diff
@@ -0,0 +1,82 @@
+  temp_coords_csv <- "final_design_points.csv"
+  file.remove(temp_ensemble_data_rds, temp_coords_csv)
+
+  set.seed(123)
```
Beware that the effects of this `set.seed` persist outside this test block! I think the easiest way to contain it would be to wrap all the lines that use random nunbers in one big `withr::with_seed(123, {...code that uses randomness here...})`.

(I think supposedly you could also replace this line with `withr::local_seed(123)` to have the seed reset at the end of the current execution context, but in my local tests that still affected code outside the function. Dunno if I'm misunderstanding "context" or I did something else wrong)

dlebauer committed:
Merge branch 'main' of github.com:pecanproject/pecan into ensemble_downscaling

dlebauer committed:
increased logging and set seed in future_map for ensemble_downscale function

dlebauer committed:
replace randomForest with ranger in ensemble_downscale for speed and parallelization

dlebauer committed:
update downscale_metrics function

dlebauer committed:
merge

dlebauer committed:
Refactor ensemble_downscale function for improved clarity and error handling

infotroph committed:
Merge branch 'develop' into ensemble_downscaling

divine7022 committed:
CH4 near zero math

divine7022 committed:
deterministic per-ensemble train/test split

dlebauer committed:
Merge pull request #10 from divine7022/ensemble_downscaling

seed the train/test split so scenarios are comparable

dlebauer committed:
Merge branch 'develop' into ensemble_downscaling

dlebauer commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -112,64 +160,82 @@ SDA_downscale_preprocess <- function(data_path, coords_path, date, carbon_pool)
   maps <- list()
   predictions <- list()
-  
+
```
```suggestion
# TODO: Consider extracting the RF and CNN branches into dedicated helper functions.
```

dlebauer committed:
Address changes requested in PR 3431

dlebauer reviewed (commented):
implemented all requested changes, and most suggestions, other than the nontrivial refactor required  to separate model fitting from spatial map prediction in https://github.com/PecanProject/pecan/pull/3431#discussion_r1940365525

dlebauer committed:
Apply suggestions from code review

Co-authored-by: David LeBauer <dlebauer@gmail.com>

dlebauer committed:
Merge branch 'develop' into ensemble_downscaling

dlebauer committed:
add `stats::` and `.data` to fix build errors

dlebauer committed:
remove duplicate ggrepel dependency

dlebauer committed:
Merge branch 'develop' into ensemble_downscaling
